dynamiqs/mesolve/rouchon.py

```diff
@@ -54,33 +82,33 @@ def Ms(self, Hnh: Tensor, fwd: bool = True) -> tuple(Tensor, Tensor):
     def T(self, R: Tensor) -> Tensor:
         # we normalize the map at each time step by inverting `R` using its Cholesky
         # decomposition `R = T @ T.mT`
-        # -> (b_H, 1, n, n)
+        # -> (b_H, b_L, 1, n, n)
         return cholesky(R)[0]  # lower triangular
 
     def forward(self, t: float, rho: Tensor) -> Tensor:
         r"""Compute $\rho(t+dt)$ using a Rouchon method of order 1.
 
         Args:
             t: Time.
-            rho: Density matrix of shape `(b_H, b_rho, n, n)`.
+            rho: Density matrix of shape `(b_H, b_L, b_rho, n, n)`.
 
         Returns:
-            Density matrix at next time step, as tensor of shape `(b_H, b_rho, n, n)`.
+            Density matrix at next time step, as tensor of shape
+            `(b_H, b_L, b_rho, n, n)`.
         """
-        # rho: (b_H, b_rho, n, n) -> (b_H, b_rho, n, n)
-
-        H = self.H(t)  # (b_H, 1, n, n)
-        Hnh = self.Hnh(H)  # (b_H, 1, n, n)
-        M0, M1s = self.Ms(Hnh)  # (b_H, 1, n, n), (1, len(L), n, n)
+        # rho: (b_H, b_L, b_rho, n, n) -> (b_H, b_L, b_rho, n, n)
+        H = self.H(t)  # (b_H, 1, 1, n, n)
+        Hnh = self.Hnh(H)  # (b_H, b_L, 1, n, n)
+        M0, M1s = self.Ms(Hnh)  # (b_H, b_L, 1, n, n), (len(L), 1, b_L, 1, n, n)
 
         # normalize the Kraus Map
         if self.options.normalize == 'cholesky':
-            R = self.R(M0)  # (b_H, 1, n, n)
-            T = self.T(R)  # (b_H, 1, n, n)
+            R = self.R(M0)  # (b_H, b_L, 1, n, n)
+            T = self.T(R)  # (b_H, b_L, 1, n, n)
             rho = inv_kraus_matmul(T.mH, rho, upper=True)  # T.mH^-1 @ rho @ T^-1
 
         # compute rho(t+dt)
-        rho = kraus_map(rho, M0) + kraus_map(rho, M1s)  # (b_H, b_rho, n, n)
+        rho = M0 @ rho @ M0.mH + kraus_map(rho, M1s)  # (b_H, b_L, b_rho, n, n)
```

Comment:
The rational for the einsum for M0 was that it's faster, let's keep this for now, I'll write smwh that we need to fix it.

Reply:
Are you sure? I'm quite skeptical that einsum is faster if we're not summing over the first dimension.

Reply:
We need to do benchmarks.